### PR TITLE
Add Lua profiler module

### PR DESCRIPTION
**Pull request by suokko**
Lua profiler module is pure lua profiler implementation. It support time based and call entry exit based profiling. The base profiler code from http://lua-users.org/wiki/PepperfishProfiler required some porting to lua 5.3 and dfhack specialization.

While I ported the code to dfhack I noticed that profiling hooks looked like doing a lot of unnecessary work. I decided to rewrite profiling hooks in hopes that lower profiling costs would improve profiling statistics quality.

### CODE REVIEW

File: docs/changelog.txt [hunk 1]
```diff
@@ -316,6 +316,9 @@ changelog.txt uses a syntax similar to RST, with a few special sequences:
 ## Structures
 - New globals: ``soul_next_id``
 
+## Lua
+- Added ``profiler`` module to measure lua performance
+
```
Comment (lethosor) [inline docs/changelog.txt]: Wrong section - this is under 0.44.05-r1.

It's possible that you based this on an old commit. If that's the case, it should be manageable enough to fix on our end.

File: library/lua/profiler.lua [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+  self.stack = {}
+  self.depth = 0
+  if self.variant == "time" then
+    self.lastclock = os.clock()
+    debug.sethook( _profiler_hook_wrapper_by_time, "", self.sampledelay )
+  elseif self.variant == "call" then
+    debug.sethook( _profiler_hook_wrapper_by_call, "cr" )
+  else
+    print("Profiler method must be 'time' or 'call'.")
+    sys.exit(1)
```
Comment (lethosor) [inline library/lua/profiler.lua]: Not acceptable in a module (but I don't think this is defined, anyway). It should use error() instead.

File: library/lua/profiler.lua [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+  if self.variant == "time" then
+    terms.capitalized = "Sample"
+    terms.single = "sample"
+    terms.pastverb = "sampled"
+  elseif self.variant == "call" then
+    terms.capitalized = "Call"
+    terms.single = "call"
+    terms.pastverb = "called"
+  else
+    assert(false)
```
Comment (lethosor) [inline library/lua/profiler.lua]: Might as well replace with `error()`, although I think the above error() should be enough.

**Comment by suokko** [comment]
Clearly I don't know how to use github. I tried to make a new profile_v2 branch with commits change to fix your comments. But then I couldn't figure out how to make this pull request reflect the changes. I found a stackoverflow answer telling I need to update my own branch and changes will automatically appear to pull requests. Too bad it seems like github doesn't like force updates.

If this pull request is broken I will open a new one.

**Comment by warmist** [comment]
@suokko I'm pretty sure you can force push to github. Just make sure that the branch was correct.

Also is this related/can be used for lua debugging/IDE integration?

**Comment by suokko** [comment]
The profiler module doesn't help integrating debugger to dfhack. Debugger requires much more calls to the debug module of lua. Writing completely new debugger or IDE integration feels like a lot work even tough the debug module seems to provide a good and simple interface.

Easiest debugging aid tool to develop based on profiler would latrace like function call tracer.

**Comment by warmist** [comment]
No, i meant this: https://studio.zerobrane.com/doc-remote-debugging

It seemed to me that we only need a functional lua-sockets?

**Comment by lethosor** [comment]
@suokko generally you should continue working on the same branch you were before, which avoids the need to force push and keeps history intact. Looks like you addressed everything except the line in the license file, which we can add pretty easily.

**Pull request merged**